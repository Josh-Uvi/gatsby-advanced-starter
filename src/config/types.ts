export interface WebsiteData {
  title: string;
  titleShort: string;
  name: string;
  description: string;
  logoUrl: string;
  fbAppId?: string;
  twitterName?: string;
  url: string;
  rss: string;
  rssTitle: string;
}

export interface UserData {
  id: string;
  firstName: string;
  lastName: string;
  twitterName?: string;
  email: string;
  location: string;
  avatar: string;
  links?: {
    label: string;
    url: string;
  }[];
}

export interface OrganizationData {
  name: string;
  description: string;
  logoUrl: string;
  url: string;
}

export interface SiteConfig {
  website: WebsiteData;
  user: UserData;
  organization: OrganizationData;
  pathPrefix: string;
  googleAnalyticsID?: string;
  disqusShortname?: string;
  postsPerPage: number;
  copyright: string;
  themeColor: string;
  backgroundColor: string;
}
