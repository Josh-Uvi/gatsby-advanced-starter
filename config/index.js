const validateSiteConfig = require("./Validator");
const userConfig = require("../data/SiteConfig.js");

module.exports = validateSiteConfig(userConfig);
