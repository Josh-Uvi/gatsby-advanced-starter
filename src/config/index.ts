import validateSiteConfig from "./Validator";
import baseConfig from "./defaultConfig";

const validatedConfig = validateSiteConfig(baseConfig);

export type {
  SiteConfig,
  OrganizationData,
  UserData,
  WebsiteData,
} from "./types";

export { default as schema } from "./schema";
export { default as constants } from "./constants";

export const defaultConfig = validatedConfig;
