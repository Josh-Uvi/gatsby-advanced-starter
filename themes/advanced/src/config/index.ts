import urlJoin from "url-join";
import { merge } from "lodash";

import validateSiteConfig from "./Validator";
import baseConfig from "./defaultConfig";
import { SiteConfig } from "./types";

const validatedConfig = validateSiteConfig(baseConfig);

export type {
  SiteConfig,
  OrganizationData,
  UserData,
  WebsiteData,
} from "./types";

export { default as schema } from "./schema";
export { default as constants } from "./constants";

export const defaultConfig = validatedConfig;

export const withDefaults = (userConfig: SiteConfig): Readonly<SiteConfig> =>
  merge(defaultConfig, userConfig);

export const withBasePath = (config: SiteConfig, url: string): string =>
  config.basePath ? urlJoin(config.basePath, url) : url;
