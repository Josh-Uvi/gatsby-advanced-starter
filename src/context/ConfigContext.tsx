import * as React from "react";
import config from "../config";

const ConfigContext = React.createContext(config);

const ConfigProvider: React.FC = ({ children }) => (
  <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
);

export default ConfigContext;
export { ConfigProvider };
