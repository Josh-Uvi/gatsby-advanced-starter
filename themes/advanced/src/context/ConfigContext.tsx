import * as React from "react";

import { useStaticQuery, graphql } from "gatsby";
import { defaultConfig, SiteConfig } from "../config";

const ConfigContext = React.createContext<SiteConfig>(defaultConfig);

type ConfigProviderProps = {
  children: React.ReactNode;
};

const ConfigProvider = ({ children }: ConfigProviderProps): JSX.Element => {
  // Query and provide SiteConfig
  const resp = useStaticQuery<GatsbyTypes.UserConfigQuery>(
    graphql`
      query UserConfig {
        site {
          siteMetadata {
            config {
              contentDir
              assetDir
              embededImageWidth
              embededVideoWidth
              basePath
              iconPath
              iconCachePath
              iconList {
                src
                sizes
                type
                purpose
              }
              organization {
                description
                logoUrl
                name
                url
              }
              pathPrefix
              user {
                about
                avatar
                firstName
                github
                email
                id
                lastName
                linkedIn
                location
                twitterName
              }
              website {
                backgroundColor
                copyright
                description
                language
                fbAppId
                googleAnalyticsId
                disqusShortname
                logoUrl
                name
                rss
                rssTitle
                themeColor
                title
                titleShort
                twitterName
                url
              }
            }
          }
        }
      }
    `
  );

  const config = resp.site?.siteMetadata?.config;

  if (!config) throw Error("ConfigProvider: Failed to query SiteConfig.");

  return (
    <ConfigContext.Provider value={config}>{children}</ConfigContext.Provider>
  );
};
export default ConfigContext;
export { ConfigProvider };
