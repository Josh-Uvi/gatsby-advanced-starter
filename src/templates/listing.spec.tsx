import React from "react";
import "jest-styled-components";
import { render } from "@testing-library/react";
import ListingPage from "./listing";
import ConfigContext from "../context/ConfigContext";
import config from "../config";

import { listingQuery } from "../../test/sampleData";

const renderPage = (postsPerPage = 3, currentPage = 1) => {
  const pageCount = Math.ceil(listingQuery.allMdx.edges.length / postsPerPage);

  const pageContext = {
    limit: postsPerPage,
    skip: (currentPage - 1) * postsPerPage,
    pageCount,
    currentPageNum: currentPage,
  };
  const data = {
    allMdx: {
      edges: listingQuery.allMdx.edges.slice(
        pageContext.skip,
        pageContext.skip + pageContext.limit
      ),
    },
  };

  const view = render(
    <ConfigContext.Provider value={{ ...config, postsPerPage }}>
      <ListingPage
        pageContext={pageContext}
        data={data as GatsbyTypes.ListingQueryQuery}
      />
    </ConfigContext.Provider>
  );

  return view.asFragment();
};

describe("page component ListingPage", () => {
  it("renders correctly with paging", () => {
    expect.assertions(3);

    // eslint-disable-next-line testing-library/render-result-naming-convention
    const firstPage = renderPage(3, 1);
    expect(firstPage).toMatchSnapshot();

    // eslint-disable-next-line testing-library/render-result-naming-convention
    const secondPage = renderPage(3, 2);
    expect(secondPage).toMatchSnapshot();

    // eslint-disable-next-line testing-library/render-result-naming-convention
    const lastPage = renderPage(3, 3);
    expect(lastPage).toMatchSnapshot();
  });

  it("renders correctly without paging", () => {
    expect.assertions(1);

    const pageContext = {
      limit: listingQuery.allMdx.edges.length,
      skip: 0,
      pageCount: 1,
      currentPageNum: 1,
    };
    const data = {
      allMdx: listingQuery.allMdx,
    };

    const { asFragment } = render(
      <ConfigContext.Provider value={{ ...config, postsPerPage: 0 }}>
        <ListingPage
          pageContext={pageContext}
          data={data as GatsbyTypes.ListingQueryQuery}
        />
      </ConfigContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
