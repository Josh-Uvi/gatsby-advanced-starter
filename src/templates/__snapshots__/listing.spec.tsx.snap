// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page component ListingPage renders correctly with paging 1`] = `
<DocumentFragment>
  .c0 {
  min-height: 100vh;
  min-width: 100vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
}

.c1 {
  min-height: 100%;
  min-width: 60%;
  margin-left: auto;
  margin-right: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media (min-width:564px) {

}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div>
          <a
            href="/big-sample-test"
          >
            <h1>
              Big Test
            </h1>
          </a>
          <a
            href="/the-fairys-witches"
          >
            <h1>
              The Fairy's Witches
            </h1>
          </a>
          <a
            href="/the-butterfly-of-the-edge"
          >
            <h1>
              The Butterfly of the Edge
            </h1>
          </a>
        </div>
      </div>
      <div
        class="paging-container"
      >
        <a
          href="/"
        >
          1
        </a>
        <a
          href="/2/"
        >
          2
        </a>
        <a
          href="/3/"
        >
          3
        </a>
        <a
          href="/2/"
        >
          Next
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`page component ListingPage renders correctly with paging 2`] = `
<DocumentFragment>
  .c0 {
  min-height: 100vh;
  min-width: 100vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
}

.c1 {
  min-height: 100%;
  min-width: 60%;
  margin-left: auto;
  margin-right: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media (min-width:564px) {

}

@media (min-width:564px) {

}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div>
          <a
            href="/random-project-4-with-super-long-title-phase-one"
          >
            <h1>
              Random project 4 With Super Long Title Phase One
            </h1>
          </a>
          <a
            href="/angels-of-mist"
          >
            <h1>
              Angels of Mist
            </h1>
          </a>
          <a
            href="/birch-in-the-roses"
          >
            <h1>
              Birch in the Roses
            </h1>
          </a>
        </div>
      </div>
      <div
        class="paging-container"
      >
        <a
          href="/"
        >
          Previous
        </a>
        <a
          href="/"
        >
          1
        </a>
        <a
          href="/2/"
        >
          2
        </a>
        <a
          href="/3/"
        >
          3
        </a>
        <a
          href="/3/"
        >
          Next
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`page component ListingPage renders correctly with paging 3`] = `
<DocumentFragment>
  .c0 {
  min-height: 100vh;
  min-width: 100vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
}

.c1 {
  min-height: 100%;
  min-width: 60%;
  margin-left: auto;
  margin-right: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media (min-width:564px) {

}

@media (min-width:564px) {

}

@media (min-width:564px) {

}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div>
          <a
            href="/the-fallen-time"
          >
            <h1>
              The Fallen Time
            </h1>
          </a>
          <a
            href="/bold-mage"
          >
            <h1>
              Bold Mage
            </h1>
          </a>
        </div>
      </div>
      <div
        class="paging-container"
      >
        <a
          href="/2/"
        >
          Previous
        </a>
        <a
          href="/"
        >
          1
        </a>
        <a
          href="/2/"
        >
          2
        </a>
        <a
          href="/3/"
        >
          3
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`page component ListingPage renders correctly without paging 1`] = `
<DocumentFragment>
  .c0 {
  min-height: 100vh;
  min-width: 100vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
}

.c1 {
  min-height: 100%;
  min-width: 60%;
  margin-left: auto;
  margin-right: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media (min-width:564px) {

}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div>
          <a
            href="/big-sample-test"
          >
            <h1>
              Big Test
            </h1>
          </a>
          <a
            href="/the-fairys-witches"
          >
            <h1>
              The Fairy's Witches
            </h1>
          </a>
          <a
            href="/the-butterfly-of-the-edge"
          >
            <h1>
              The Butterfly of the Edge
            </h1>
          </a>
          <a
            href="/random-project-4-with-super-long-title-phase-one"
          >
            <h1>
              Random project 4 With Super Long Title Phase One
            </h1>
          </a>
          <a
            href="/angels-of-mist"
          >
            <h1>
              Angels of Mist
            </h1>
          </a>
          <a
            href="/birch-in-the-roses"
          >
            <h1>
              Birch in the Roses
            </h1>
          </a>
          <a
            href="/the-fallen-time"
          >
            <h1>
              The Fallen Time
            </h1>
          </a>
          <a
            href="/bold-mage"
          >
            <h1>
              Bold Mage
            </h1>
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
